File: 383_RansomNote.py
# 39.49%
import collections
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        mCounter = collections.Counter(magazine)
        #rCounter = collections.Counter(ransomNote)
        for c in ransomNote:
            if c not in mCounter or mCounter[c] == 0:
                return False
            mCounter[c] -= 1
        return True

# Wonderful use on python		
def canConstruct(self, ransomNote, magazine):
    return not collections.Counter(ransomNote) - collections.Counter(magazine)
	
	#or use it in this direct way
	c = Counter(magazine)
    c.subtract(Counter(ransomNote))
    return all(v >= 0 for v in c.values())

File: 383_RansomNote.cpp
//35.93%
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> m;
		for(char c : magazine)
			m[c]++;
		for(char c : ransomNote)
			if(m.find(c) == m.end() || m[c]-- == 0)
				return false;
		return true;
    }
};


class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> map(26);
        for (int i = 0; i < magazine.size(); ++i)
            ++map[magazine[i]];
        for (int j = 0; j < ransomNote.size(); ++j)
            if (--map[ransomNote[j]] < 0)
                return false;
        return true;
    }
};

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        vector<int> vec(26, 0);
        for (int i = 0; i < magazine.size(); ++i)
            ++vec[magazine[i] - 'a'];
        for (int j = 0; j < ransomNote.size(); ++j)
            if (--vec[ransomNote[j] - 'a'] < 0)
                return false;
        return true;
    }
};